src/coreclr/debug/daccess/request_svr.cpp

```diff
@@ -99,33 +99,57 @@ HRESULT ClrDataAccess::GetServerAllocData(unsigned int count, struct DacpGenerat
     return S_OK;
 }
 
-HRESULT ClrDataAccess::ServerGCHeapDetails(CLRDATA_ADDRESS heapAddr, DacpGcHeapDetails *detailsData)
+HRESULT
+ClrDataAccess::ServerGCHeapDetails(CLRDATA_ADDRESS heapAddr, DacpGcHeapDetails *detailsData)
 {
+    // Make sure ClrDataAccess::GetGCHeapStaticData() is updated as well.
     if (!heapAddr)
     {
         // PREfix.
         return E_INVALIDARG;
     }
+    if (detailsData == NULL)
+    {
+        return E_INVALIDARG;
+    }
 
-    DPTR(dac_gc_heap) pHeap = __DPtr<dac_gc_heap>(TO_TADDR(heapAddr));
-    int i;
+    TADDR heapAddress = TO_TADDR(heapAddr);
+    dac_gc_heap heap = LoadGcHeapData(heapAddress);
+    dac_gc_heap* pHeap = &heap;
 
     //get global information first
     detailsData->heapAddr = heapAddr;
 
     detailsData->lowest_address = PTR_CDADDR(g_lowest_address);
     detailsData->highest_address = PTR_CDADDR(g_highest_address);
-    detailsData->card_table = PTR_CDADDR(g_card_table);
+    detailsData->current_c_gc_state = (CLRDATA_ADDRESS)*g_gcDacGlobals->current_c_gc_state;
 
     // now get information specific to this heap (server mode gives us several heaps; we're getting
     // information about only one of them.
     detailsData->alloc_allocated = (CLRDATA_ADDRESS)pHeap->alloc_allocated;
     detailsData->ephemeral_heap_segment = (CLRDATA_ADDRESS)dac_cast<TADDR>(pHeap->ephemeral_heap_segment);
+    detailsData->card_table = (CLRDATA_ADDRESS)pHeap->card_table;
+    detailsData->mark_array = (CLRDATA_ADDRESS)pHeap->mark_array;
+    detailsData->next_sweep_obj = (CLRDATA_ADDRESS)pHeap->next_sweep_obj;
+    if (pHeap->saved_sweep_ephemeral_seg.IsValid())
+    {
+        detailsData->saved_sweep_ephemeral_seg = (CLRDATA_ADDRESS)dac_cast<TADDR>(pHeap->saved_sweep_ephemeral_seg);
+        detailsData->saved_sweep_ephemeral_start = (CLRDATA_ADDRESS)*pHeap->saved_sweep_ephemeral_start;
+    }
+    else
+    {
+        // with regions, we don't have these variables anymore
+        // use special value -1 in saved_sweep_ephemeral_seg to signal the region case
+        detailsData->saved_sweep_ephemeral_seg = (CLRDATA_ADDRESS)-1;
+        detailsData->saved_sweep_ephemeral_start = 0;
+    }
+    detailsData->background_saved_lowest_address = (CLRDATA_ADDRESS)pHeap->background_saved_lowest_address;
+    detailsData->background_saved_highest_address = (CLRDATA_ADDRESS)pHeap->background_saved_highest_address;
 
     // get bounds for the different generations
-    for (i=0; i<NUMBERGENERATIONS; i++)
+    for (unsigned int i=0; i < *g_gcDacGlobals->max_gen + 2; i++)
```

Comment:
The DacpGcHeapDetails generation_table and finalization_fill_pointers are fixed size arrays with sizes defined as DAC_NUMBERGENERATIONS and DAC_NUMBERGENERATIONS+3 respectively. We shouldn't be using dynamically computed bounds to fill a statically sized buffer. https://github.com/dotnet/runtime/blob/main/src/coreclr/inc/dacprivate.h#L738

You could return an error HRESULT if the number of generations determined dynamically doesn't match the static sizes but we wouldn't want to overflow the buffer. A more extensive change would be to modify the DAC<->SOS interface to add support for dynamically determined numbers of generations.